README
xml-api
=======

  XML::API is a class for creating XML documents using object method
  calls. This class is meant for generating XML programatically and not
  for reading or parsing it.

INSTALLATION

  To install this module type the following:

     perl Makefile.PL
     make
     make test
     make install

DEPENDENCIES

     XML::Parser::Expat      (from XML::Parser on CPAN)

COPYRIGHT AND LICENCE

  Copyright (C) 2004,2005 Mark Lawrence <nomad@null.net>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.



Changes
- testing: Handle '\r\n' vs '\n' issues on Windows/Dos based filesystems.
- Makefile.PL: Remove all dependencies on Test::Pod* modules
- Remove auto_install from Makefile.PL
- Convert "UNIVERSAL::isa(" to "eval { $obj->isa("
- Fix object stringification tests on recent Test::More.
- Add respository information to the META data.
- Make _add() actually croak/carp on bad input.
- Update MANIFEST with Module::Install changes.
- CPAN - Released as version 0.25.


example/doc.pl
#!/usr/bin/perl
# ----------------------------------------------------------------------
# Copyright (C) 2005 Mark Lawrence <nomad@null.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# ----------------------------------------------------------------------
# Parse HTML into Perl/XML::API Code
# ----------------------------------------------------------------------

use lib 'lib'; use lib '../lib';
use XML::API;
use Storable qw(freeze thaw);
use Data::Dumper;

if (@ARGV < 1) {
    print STDERR "usage: $0 <file> [tag1 tag2 ...]\n";
    exit 1;
}

my $file = shift;

if ($file eq '-') {
    *FH = *STDIN;
}
elsif (!open(FH, $file)) {
    print STDERR "Could not open $file: $!\n";
    exit 1;
}

my $x = XML::API->new(doctype => 'xhtml');

$x->_parse_allow_tags(@ARGV);
$x->_parse(<FH>);

$x->_print();


__END__

=head1 NAME

XML::API - Perl extension for creating XML documents

=head1 SYNOPSIS

As a simple example the following perl code:

  use XML::API;
  my $x = XML::API->new(doctype => 'xhtml');
  
  $x->head_open();
  $x->title('Test Page');
  $x->head_close();

  $x->body_open();
  $x->div_open({id => 'content'});
  $x->p('A test paragraph');
  $x->div_close();
  $x->body_close();

  $x->_print;

will produce the following nicely rendered output:
  
  <?xml version="1.0" encoding="ISO-8859-1"?>
  <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
      "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
  <html>
    <head>
      <title>Test Page</title>
    </head>
    <body>
      <div id="content">
        <p>A test paragraph</p>
      </div>
    </body>
  </html>

There are also more powerful and flexible ways to use this module. Read on.

=head1 DESCRIPTION

B<XML::API> is a class for creating XML documents using
object method calls. This class is meant for generating XML
programatically and not for reading or parsing it.

The methods of a B<XML::API> object are derived directly from the XML
Schema Definition document for the desired document type.
A document author calls the desired methods (representing elements) to
create an XML tree in memory which can then be rendered or saved as desired.
The advantage of having the in-memory tree is that you can be very flexible
about when different parts of the document are created and the final output
is always nicely rendered.

It is also worth having a small through about future proofing: Because
your 'xml' is actually 'code' it is potentially possible to change the
output of *all* of your xml by modifying the Perl class. No more hunting
through source files changing strings when the next version of xhtml
comes out.

=head1 TUTORIAL

The first step is to create an object. The 'doctype' attribute determines
the XSD to use (currently only xhtml and rss are distributed with the
distribution):

  use XML::API;
  my $x = XML::API->new(doctype => 'xhtml');

$x is the only object we need for our entire XHTML document. By default
$x consists initially of only the root element ('html') which should be
thought of as the 'current' or 'containing' element. The next step might
be to add a 'head' element. We do this by calling the head_open() method:

  $x->head_open();

Because we have called a *_open() function the 'current' or 'containing'
element is now 'head'. All further elements will be added inside the
'head' element. So lets add a title element and the title content
('Document Title') to our object:

  $x->title('Document Title');

The 'title()' method on its own (ie not 'title_open()') indicates that we
are specifiying the entire title element. Further method calls will
continue to place elements inside the 'head' element until we specifiy we
want to move on by calling the _close method:

  $x->head_close();

This sets the current element back to 'html'.

So, basic elements seem relatively easy. How do we create elements with
attributes? When either the element() or element_open() methods are called
with a hashref argument the keys and values of the hashref become the
attributes:

  $x->body_open({id => 'bodyid'}, 'Content', 'more content');

By the way, both the element() and element_open() methods take arbitrary
numbers of content arguments as shown above. However if you don't want to
specify the content of the element at the time you open it up you can
use the _add() utility method later on:

  $x->div_open();
  $x->_add('Content added after the _open');

The final thing is to close out the elements and render the docment.

  $x->div_close();
  $x->body_close();
  print $x->_as_string();

Because we are not adding any more elements or content it is not strictly
necessary to close out all elements, but consider it good practice.

=head1 METHODS

=head2 new(doctype => '(xhtml|rss)', [ element => 'xxx', strict => bool ])

Create a new XML::API based object. What you get back is actually
an object of type XML::API::<doctype> which is derived from XML::API.
The containing or 'current' element is by default the root element of
the XSD.

You don't have to start with the root element of the specification. If
you wanted for instance to create a standalone 'div' object you can
with the 'element' attribute to the new() call:

  my $div = XML::API->new(doctype => 'xhtml', element => 'div');

You can then add this element to an object that has the root element 
using the _add() method below.

By default strict checking is performed to make sure that the structure
of the document matches the Schema. This can be turned off by setting
'strict' to false (0 or undef).

=head2 $x->element_open({attribute => $value}, $content)

Add a new element to the 'current' element, and set the current element
to be the element just created. Returns a reference (private data type)
to the new element which can be used in the _goto function below.

Ie given that $x currently represents:

  <html>  <---- 'current' element
          <---- future elements/content goes here
  </html>

then $x->head_open({attribute => $value}) means the tree is now:

  <html>
    <head attribute="$value">  <---- 'current' element
                               <---- future elements/content goes here
    </head>
  </html>

=head2 $x->element({attribute => $value}, $content)

Add a new element to the 'current' element but keep the 'current'
element the same. Returns a reference (private data type)
to the new element which can be used in the _goto function below.

Ie given that $x is currently:

  <div>  <---- 'current' element
         <---- future elements/content goes here
  </div>

then $x->p({attribute => $value}, $content) means the tree is now:

  <div>                      <---- still 'current' element
    <p attribute="$value">$content</p>
                             <---- future elements/content goes here
  </div>

If $content is not given (or not added with the _add method) then the
element will be rendered as empty. Eg $x->br() produces:

  <div>                      <---- still 'current' element
    <p attribute="$value">$content</p>
    <br />
                             <---- future elements/content goes here
  </div>

=head2 $x->element_close( )

This does not actually modify the tree but simply tells the object that
future elements will be added to the parent of the current element.
Ie given that $x currently represents:

  <div>
    <p>  <---- 'current' element
      $content
           <---- future elements/content goes here
    </p>
  </div>

then $x->p_close() means the tree is now:

  <div>    <---- 'current' element
    <p>
      $content
    </p>
           <---- future elements go here
  </div>

If you try to call a _close() method that doesn't match the current
element a warning will be issued and the call will fail.

=head2 $x->_add($content)

Adds content to the 'current' element. $content can be either scalar
(string, numeric) or an XML::API element.

=head2 $x->_current( )

Returns a reference (private data type) to the current element.

=head2 $x->_goto($reference)

Change the 'current' element. $reference is the return value of one of
the element_open(), element() or _current() methods.

This is useful if you create the basic structure of your document, but
then later want to go back and modify it or fill in the details.

This method also currently takes a string matching the id= attribute
of an element previously defined, but as this is specific to xhtml
don't rely on it always being around.

=head2 $x->_attrs( )

Allows you to get/set the attributes of the current element. Accepts
and returns and hashref.

=head2 $x->_comment($comment)

Is a shortcut for $x->_add('\n<!--', $content, '-->')

=head2 $x->_cdata( )

Is a shortcut for $x->_add("\n<![CDATA[", @_, " ]]>");

=head2 $x->_as_string( )

Returns the rendered version of the XML document.

=head2 $x->_print( )

A shortcut for "print $x->_as_string()"

=head1 SEE ALSO

L<XML::API::XHTML::Parse>

B<XML::API> was written for the Rekudos framework:  http://rekudos.net/

=head1 AUTHOR

Mark Lawrence E<lt>nomad@null.netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2004,2005 Mark Lawrence <nomad@null.net>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

=cut



lib/XML/API/RSS.pm
package XML::API::RSS;
# ----------------------------------------------------------------------
# Copyright (C) 2004 Mark Lawrence <nomad@null.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# ----------------------------------------------------------------------
# Derivative of XML::API for RSS 
# ----------------------------------------------------------------------
use strict;
use warnings;
use 5.006;
use base qw(XML::API);

our $VERSION = $XML::API::VERSION;

my $xsd = {};

sub _doctype {
    return '';
}

sub _xsd {
    return $xsd;
}

sub _root_element {
    return 'rss';
}

sub _root_attrs {
    return {version => '2.0'};
}


1;

__END__


=head1 NAME

XML::API::XHTML - XHTML generation through an object API

=head1 SYNOPSIS

As a simple example the following perl code:

  use XML::API::XHTML;
  my $x = new XML::API::XHTML();
  
  $x->head_open();
  $x->title('Test Page');
  $x->head_close();

  $x->body_open();
  $x->div_open({id => 'content'});
  $x->p('A test paragraph');
  $x->div_close();
  $x->body_close();

  $x->_print;

will produce the following nicely rendered output:
  
  <?xml version="1.0" encoding="ISO-8859-1"?>
  <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
      "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
  <html>
    <head>
      <title>Test Page</title>
    </head>
    <body>
      <div id="content">
        <p>A test paragraph</p>
      </div>
    </body>
  </html>

There are more complicated and flexible ways to use this module. Read on.

=head1 DESCRIPTION

B<XML::API::XHTML> is a perl object class for creating XHTML documents. The
methods of a B<XML::API::XHTML> object are derived directly from the XHTML
specification. A document author uses the methods to define the structure
and content of their document, which can then be printed and sent somewhere
or saved as desired.

The first step is to create an object:

  use XML::API::XHTML;
  my $x = new XML::API::XHTML();

$x is the only object we need for our entire XHTML document. By default
$x consists initially of only the root element ('html') which should be
thought of as the 'current' or 'containing' element. The next step might
be to add a 'head' element. We do this by calling the head_open() method:

  $x->head_open();

Because we have called a *_open() function all further elements will be
added inside the 'head' element. So lets add the title element and the title
itself ('Document Title') to our object:

  $x->title('Document Title');

The 'title()' method on its own (ie not 'title_open()') indicates that we
are finished with the title element. Further methods will
continue to place elements inside the 'head' element until we specifiy
we want to move on by calling the _close method:

  $x->head_close();

We are now back inside the 'html' element.

So, basic elements seem relatively easy. How do we create elements with
attributes? When either the element() or element_open() methods are called
with a hashref argument the keys and values of the hashref become the
attributes:

  $x->body_open({id => 'bodyid'}, 'Content', 'more content');

By the way, both the element() and element_open() methods take arbitrary
numbers of content arguments as shown above. However if you don't want to
specify the content of the element at the time you open it up you can
use the _add() utility method later on:

  $x->div_open();
  $x->_add('Content added afterwards');

The final thing is to close out the elements and render our docment.
It is not strictly necessary to close out all elements, but consider it
good practice.

  $x->div_close();
  $x->body_close();
  print $x->_as_string();

=head1 SCHEMA CONFORMANCE

If you attempt to call a method that would create a document that doesn't
conform to the schema an error will be printed to STDERR and the method
will fail.

=head1 METHODS

B<XML::API::XHTML> inherits directly from B<XML::API> and actually provides
no methods of its own. All utility methods (new(), _print(), _add etc)
and advanced functionality such as adding content in a non-linear fashion
are base properties of B<XML::API> and users are directed there
for details. The important B<XML::API> methods are only summarised
here for convenience.

=head2 B<new()>

All elements of the XHTML specification (with the exception of 'html')
are available as the following function forms:

  $x->element_open({attribute => $value}, $content)

  $x->element({attribute => $value}, $content)

  $x->element_close({attribute => $value}, $content)


This module has no control functions of its own. See the documentation for
L<XML::API> for useful functions (such as _print).


=head2 EXPORT

None.

=head2 REQUIRES

Perl5.6.0, Carp, XML::API

=head1 SEE ALSO

XML::API and XML::API::XHTML were both written
for the Rekudos framework which is housed at http://rekudos.net/.

=head1 AUTHOR

Mark Lawrence E<lt>nomad@null.net<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2004 Mark Lawrence <nomad@null.net>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

=cut



lib/XML/API/XHTML.pm
package XML::API::XHTML;
# ----------------------------------------------------------------------
# Copyright (C) 2004 Mark Lawrence <nomad@null.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# ----------------------------------------------------------------------
# Derivative of XML::API for XHTML 
# ----------------------------------------------------------------------
use strict;
use warnings;
use base qw(XML::API);

our $VERSION = '0.13';

use constant DOCTYPE => qq{<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">};

use constant XSD => {};
use constant ROOT_ELEMENT => 'html';
use constant ROOT_ATTRS => {xmlns => 'http://www.w3.org/1999/xhtml'};


my $xsd = {};

sub _doctype {
    return q{<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">};
}

sub _xsd {
    return $xsd;
}

sub _root_element {
    return 'html';
}

sub _root_attrs {
    return {xmlns => 'http://www.w3.org/1999/xhtml'};
}


1;

__END__


=head1 NAME

XML::API::XHTML - XHTML generation through an object API

=head1 SYNOPSIS

As a simple example the following perl code:

  use XML::API::XHTML;
  my $x = new XML::API::XHTML();
  
  $x->head_open();
  $x->title('Test Page');
  $x->head_close();

  $x->body_open();
  $x->div_open({id => 'content'});
  $x->p('A test paragraph');
  $x->div_close();
  $x->body_close();

  $x->_print;

will produce the following nicely rendered output:
  
  <?xml version="1.0" encoding="ISO-8859-1"?>
  <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
      "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
  <html>
    <head>
      <title>Test Page</title>
    </head>
    <body>
      <div id="content">
        <p>A test paragraph</p>
      </div>
    </body>
  </html>

There are more complicated and flexible ways to use this module. Read on.

=head1 DESCRIPTION

B<XML::API::XHTML> is a perl object class for creating XHTML documents. The
methods of a B<XML::API::XHTML> object are derived directly from the XHTML
specification. A document author uses the methods to define the structure
and content of their document, which can then be printed and sent somewhere
or saved as desired.

The first step is to create an object:

  use XML::API::XHTML;
  my $x = new XML::API::XHTML();

$x is the only object we need for our entire XHTML document. By default
$x consists initially of only the root element ('html') which should be
thought of as the 'current' or 'containing' element. The next step might
be to add a 'head' element. We do this by calling the head_open() method:

  $x->head_open();

Because we have called a *_open() function all further elements will be
added inside the 'head' element. So lets add the title element and the title
itself ('Document Title') to our object:

  $x->title('Document Title');

The 'title()' method on its own (ie not 'title_open()') indicates that we
are finished with the title element. Further methods will
continue to place elements inside the 'head' element until we specifiy
we want to move on by calling the _close method:

  $x->head_close();

We are now back inside the 'html' element.

So, basic elements seem relatively easy. How do we create elements with
attributes? When either the element() or element_open() methods are called
with a hashref argument the keys and values of the hashref become the
attributes:

  $x->body_open({id => 'bodyid'}, 'Content', 'more content');

By the way, both the element() and element_open() methods take arbitrary
numbers of content arguments as shown above. However if you don't want to
specify the content of the element at the time you open it up you can
use the _add() utility method later on:

  $x->div_open();
  $x->_add('Content added afterwards');

The final thing is to close out the elements and render our docment.
It is not strictly necessary to close out all elements, but consider it
good practice.

  $x->div_close();
  $x->body_close();
  print $x->_as_string();

=head1 SCHEMA CONFORMANCE

If you attempt to call a method that would create a document that doesn't
conform to the schema an error will be printed to STDERR and the method
will fail.

=head1 METHODS

B<XML::API::XHTML> inherits directly from B<XML::API> and actually provides
no methods of its own. All utility methods (new(), _print(), _add etc)
and advanced functionality such as adding content in a non-linear fashion
are base properties of B<XML::API> and users are directed there
for details. The important B<XML::API> methods are only summarised
here for convenience.

=head2 B<new()>

All elements of the XHTML specification (with the exception of 'html')
are available as the following function forms:

  $x->element_open({attribute => $value}, $content)

  $x->element({attribute => $value}, $content)

  $x->element_close({attribute => $value}, $content)


This module has no control functions of its own. See the documentation for
L<XML::API> for useful functions (such as _print).


=head2 EXPORT

None.

=head2 REQUIRES

Perl5.6.0, Carp, XML::API

=head1 SEE ALSO

XML::API and XML::API::XHTML were both written
for the Rekudos framework which is housed at http://rekudos.net/.

=head1 AUTHOR

Mark Lawrence E<lt>nomad@null.net<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2004 Mark Lawrence <nomad@null.net>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

=cut



inc/Module/Install/MakeMaker.pm
#line 1 "inc/Module/Install/MakeMaker.pm - /usr/share/perl5/Module/Install/MakeMaker.pm"
# $File: //depot/cpan/Module-Install/lib/Module/Install/MakeMaker.pm $ $Author: autrijus $
# $Revision: #11 $ $Change: 1781 $ $DateTime: 2003/10/22 17:14:03 $ vim: expandtab shiftwidth=4

package Module::Install::MakeMaker;
use Module::Install::Base; @ISA = qw(Module::Install::Base);

$VERSION = '0.01';

use ExtUtils::MakeMaker ();

my $makefile;
sub WriteMakefile {
    my ($self, %args) = @_;
    $makefile = $self->load('Makefile');

    # mapping between MakeMaker and META.yml keys
    $args{MODULE_NAME} = $args{NAME};
    unless ($args{NAME} = $args{DISTNAME} or !$args{MODULE_NAME}) {
        $args{NAME} = $args{MODULE_NAME};
        $args{NAME} =~ s/::/-/g;
    }

    foreach my $key (qw(name module_name version version_from abstract author)) {
        my $value = delete($args{uc($key)}) or next;
        $self->$key($value);
    }

    if (my $prereq = delete($args{PREREQ_PM})) {
        while (my($k,$v) = each %$prereq) {
            $self->requires($k,$v);
        }
    }

    # put the remaining args to makemaker_args
    $self->makemaker_args(%args);
}

END {
    if ($makefile) {
        $makefile->write;
        $makefile->Meta->write;
    }
}

1;


Makefile.PL
use 5.006;
use inc::Module::Install;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

my $version = '0.13';
my $name    = 'XML::API';
my $fname   = "XML-API-$version";

my $prereqs = {
    'Test::More'         => 0,
    'Tree::Simple'       => 1.17,
    'XML::Parser::Expat' => 0,
};

sub MY::postamble { "

changes:
	echo 'Changelog for Perl extension $name' > $$
	echo '' >> $$
	-darcs changes >> $$ && mv $$ Changes
	rm -f $$
                                
deb: changes
	(head -1 debian/changelog | grep $version) || exit 1
	rm -f $fname.tar.gz *.deb *.changes
	-chmod 755 debian/rules
	\$(MAKE) dist
	tar zxvf $fname.tar.gz
	cd $fname && dpkg-buildpackage -b -rfakeroot -tc

deba:
	sudo cp *.deb *.changes /var/lib/debarchiver/incoming
	sudo debarchiver -so

tgz:
	rm -f XML-API.tgz
	ln -s $fname.tar.gz XML-API.tgz

";
}

WriteMakefile(
    NAME              => $name,
    VERSION           => $version,
    PREREQ_PM         => $prereqs,
    ABSTRACT          => 'Perl extension for creating XML documents',
    AUTHOR            => 'Mark Lawrence <nomad@null.net>',
);
